File: 345. Reverse Vowels of a String/reverse_vowels.js
/**
    * Solved By: Gabriel E. Rodriguez Garcia
    * Date Completed: July 31, 2023
    * Exersise #345
    * Exercise Name: Reverse Vowels of a String
    * Difficulty: Easy
    * Time Complexity: O(n)
    * Space Complexity: O(n)
    * Study Plan: LeetCode 75
    * Category: Array / String 
    * Language Used: JavaScript
    * Tags: Two Pointer, String
 */


/**
 * Reverses all vowels found in a string
 * @param {string} s A String containing a vowels to reverse
 * @return {string} A string with the vowels reversed.
 */
var reverseVowels = function(s) {
    //Staring index
    var start = 0;
    //Ending index
    var end = s.length-1;
    //String with all vowels.
    var vowels = "aeiouAEIOU";
    //Array filled with each character in the word
    var letters = s.split("");

    //Traverse the array from the start and the end at the same time
    while(start<end) {
        //If the letter isnt a vowel move to the next letter
        if(!vowels.includes(letters[start]))
            start++;
        //If the letter is not a vowel move to the next letter
        if(!vowels.includes(letters[end]))
            end--;
        //If both current letters are vowels then swap them
        if(vowels.includes(letters[start]) && vowels.includes(letters[end])){
            //Store the start variable in a temporary variable
            var temp = letters[start];
            //Replace the start with the end vowel
            letters[start] = letters[end];
            //Replace the end with the temporary vowel
            letters[end] = temp;
            //Move to the next letter
            start++;
            //Move to the next letter
            end--;
        }

    }
    //Returns the joined string of the letters array
     return letters.join("");

    
};

File: 345. Reverse Vowels of a String/reverse_vowels.py
"""
    * Solved By: Gabriel E. Rodriguez Garcia
    * Date Completed: July 31, 2023
    * Exersise #345
    * Exercise Name: Reverse Vowels of a String
    * Difficulty: Easy
    * Time Complexity: O(n)
    * Space Complexity: O(n)
    * Study Plan: LeetCode 75
    * Category: Array / String 
    * Language Used: Python
    * Tags: Two Pointer, String
"""


"""
    * Reverses all vowels found in a string
    * @param s A String containing a vowels to reverse
    * @return A string with the vowels reversed.
"""
def reverseVowels(s: str) -> str:
        #Start Index
        start = 0
        #End Index
        end = len(s)-1
        #String containing all vowels
        vowels = "aeiouAEIOU"
        #List of letters in the string
        letters = list(s)

        #Traverses the array from both the start and end
        while(start<end):
            #If the letter is not a vowel move to the next letter
            if(letters[start] not in vowels):
                start +=1
            #IF the letter is not a vowel move to the next letter
            if(letters[end] not in vowels):
                end -=1
            #If both letters are vowels 
            if(letters[start] in vowels and letters[end] in vowels):
                #fill the temp with the start vowel
                temp= letters[start]

                #Store the end vowel in the start
                letters[start] = letters[end]

                #Store the start vowel in the end
                letters[end] = temp

                #Move to the next letter
                start+=1
                #Move to the next letter
                end-=1

        #Join the new letters in a string
        return "".join(letters)